kiwmi/desktop/layer_shell.c

```diff
@@ -12,26 +12,65 @@
 #include <pixman.h>
 #include <wayland-server.h>
 #include <wlr/types/wlr_layer_shell_v1.h>
+#include <wlr/types/wlr_output_layout.h>
 #include <wlr/util/log.h>
 
 #include "desktop/desktop.h"
 #include "desktop/output.h"
+#include "desktop/stratum.h"
+#include "desktop/toplevel.h"
 #include "input/seat.h"
 #include "server.h"
 
+static void
+focus_topmost_layer(struct kiwmi_output *output)
+{
+    uint32_t layers_above_shell[] = {
+        ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY,
+        ZWLR_LAYER_SHELL_V1_LAYER_TOP,
+    };
+    size_t nlayers = sizeof(layers_above_shell) / sizeof(layers_above_shell[0]);
+    struct kiwmi_layer *layer;
+    struct kiwmi_layer *topmost = NULL;
+    for (size_t i = 0; i < nlayers; ++i) {
+        wl_list_for_each_reverse (
+            layer, &output->layers[layers_above_shell[i]], link) {
+            if (layer->layer_surface->current.keyboard_interactive) {
+                topmost = layer;
+                break;
+            }
+        }
+
+        if (topmost) {
+            break;
+        }
+    }
+
+    if (topmost) {
+        wlr_scene_node_raise_to_top(&topmost->toplevel.scene_subtree->node);
+        wlr_scene_node_raise_to_top(
+            &topmost->toplevel.popups_scene_subtree->node);
+    }
+
+    struct kiwmi_server *server =
+        wl_container_of(output->desktop, server, desktop);
+    seat_focus_layer(server->input.seat, topmost);
+}
+
 static void
 kiwmi_layer_destroy_notify(struct wl_listener *listener, void *UNUSED(data))
 {
     struct kiwmi_layer *layer = wl_container_of(listener, layer, destroy);
 
+    wlr_scene_node_destroy(&layer->toplevel.scene_subtree->node);
+    wlr_scene_node_destroy(&layer->toplevel.popups_scene_subtree->node);
+
     wl_list_remove(&layer->link);
     wl_list_remove(&layer->destroy.link);
     wl_list_remove(&layer->map.link);
     wl_list_remove(&layer->unmap.link);
 
-    wlr_layer_surface_v1_close(layer->layer_surface);
-
-    arrange_layers(layer->output);
```

Comment:
self-note: this can apparently be of use when waybar is ctrl-c'd (with this removed, the exclusive zone didn’t disappear)